src/testing/flow_solver.cpp

```diff
@@ -372,7 +410,7 @@ int FlowSolver<dim,nstate>::run_test() const
         // Time advancement loop with on-the-fly post-processing
         //----------------------------------------------------
         pcout << "Advancing solution in time... " << std::endl;
-        while(ode_solver->current_time < final_time)
+        while((final_time - ode_solver->current_time) > (0.5 * constant_time_step)) //comparing to constant_time_step in case roundoff errors exceed machine zero
```

Comment:
Instead of a while loop, would it make sense to somehow make sure that the last step lands exactly on the desired final time? Like if we want to go up to final_time = 1.0, but with constant time step of 0.3, we'd do
t0 = 0.0
t1 = t0 + 0.3 = 0.3
t2 = t1 + 0.3 = 0.6
t3 = t2 + 0.3 = 0.9
**t3 = t3 + 0.1 = 1.0**

Something like
```cpp
int n_steps = ((final_time - initial_time) / constant_step) // Rounds down to nearest integer
const time_remainder = final_time - n_steps * constant_time_step;
for (int i_step = 0; i_step < n_steps +1; ++i_step) 
{ 
    do work 
}
advance_solution_time(time_remainder);

```

@jbrillon

Reply:
Otherwise, we might end up comparing various solutions at different final times, which can easily mess up convergence orders.

Reply:
OK, but the comment above still holds. What happens if constant_time_step is not an integer multiple of final time?

Reply:
For unsteady simulations that use a CFL number to compute the dt or `constant_time_step`, the final time is most likely not going to be exactly equal to the final time you see on the plots in papers

Reply:
OK, but you would want to have the final time of the simulation to be the final time you describe in your paper... Need something such that if (current_time + time_step_evaluated_by_CFL_or_given_by_constant_time_step) is greater than (final_time), take a time_step = (final_time - current_time)

Reply:
If you're using a CFL, I think its understood that when you write final time of the simulation, the actual final time could be off by a very small amount say 1e-4 or even smaller than that. Also, the current implementation ensures that we reach the _at least_ final time since it checks `current_time<final_time` then does `step_in_time` (which updates the current time) so the actual final time >= final time

Reply:
Sure, so... you might be off up to O(dt)... Aka.. you're back to a Euler-explicit scheme accuracy...

Reply:
Okay sure, if I was comparing the solution at the final time then yes, but in that case, I would modify the CFL to be slightly different than the inputted one so that I exactly reach the inputted final time (done before advancing in time), I could introduce a function called `modify_time_step_to_exactly_reach_desired_final_time` and call that after doing `get_constant_time_step`; I can open a quick PR for this. What do you think?

Reply:
I made this change because I'm using a `dt` which should evenly divide `final_time`. However, sometimes the computation would run until (`final_time + dt`) rather than stopping at `final_time`. The while loop entered because `current_time` was on the order of 1E-14 less than `final_time`. My fix compares to `dt` such that machine errors like this don't cause the flow solver to overshoot `final_time`. I don't have a particularly good reason for comparing to `0.5*dt` rather than any other constant of the same order.

I could revert this change and compare to a different `final_time_actual` (or something like that). However, I don't think this is the best solution going forward. Comparing at a different `final_time_actual` doesn't make a huge difference for advection. but for a more complicated PDE where the solution evolves between `final_time` and `final_time_actual`, I am not confident in the validity of comparing convergence at a different point in the solution. 

So, I'm not sure that my solution is the best one, but I think it's worth looking into a better way to handle this in the future.

Reply:
As a fix for this problem, my latest commit compares to 1E-13 rather than 0.5 dt to avoid taking an extra step. My test runs as before.

In the future, could look into making the tolerance scale with dt if this issue resurfaces, but this fix works with my test and should not impact the results of other tests.

Reply:
Mmmmh, I still think that something like what I mentionned above is needed. I don't think you want the code to modify the input timestep or CFL. Since the code itself is a research code and we end up doing temporal studies, you'd want the specified timestep to be respected.
The solution I proposed with taking as many steps as you can until you need to reduce the last one is still my preferred solution.

This one:
t0 = 0.0
t1 = t0 + 0.3 = 0.3
t2 = t1 + 0.3 = 0.6
t3 = t2 + 0.3 = 0.9
**t4 = t3 + 0.1 = 1.0**

In effect, this would have a convergence of order O(0.3^p) even if the last step is 0.1, because most of the steps would be 0.3. If you decide to modify the stepping scheme, you would end up modifying it to 0.25, which wouldn't give the correct orders if you assumed it was halved.

Reply:
@dougshidong I agree that your solution should preserve the order, because the remainder step is always smaller than the target dt, so error should still be dominated by (target_dt)^p. However, it's probably more sound for the time refinement studies to always use an evenly-divisible dt when doing temporal convergence studies. If the step size is calculated by a CFL number, then the final time could be adjusted at the first refinement (e.g. with your example, stop all computations at 0.9).

To address this, I will add an error check to the time refinement study that makes sure that `final_time` is divisible by `initial_time_step`. This will ensure that such an issue isn't encountered for a constant-time-step test like the one I have added.

Another complication to this discussion is the assumption that we know the step size a priori. I think that the solution I implemented a few commits ago will give me problems when I add relaxation RK timestepping. In that type of timestepping, the timestep taken at each step varies like dt_n = (1 + target_dt^(p-1) ) * target_dt. In that case, the 1E-13 tolerance is not guaranteed to capture the desired final time if it happens that the variations are on the lower side. @dougshidong's solution doesn't quite allow for this because it requires that we know `time_remainder` exactly. This isn't immediately necessary to be addressed, but an issue that I am anticipating for the future.

Reply:
I think it's worth opening up an issue about it since we do care about OOA in a high order code like PHiLiP and it is fair that an unknowing student would assume that the final time specified actually is the final time reached.

Reply:
Ok, how about just outputting a warning if `mod(final_time,constant_time_step)!=0` ?